### PR TITLE
GC Clamp Property Added to Primer Class

### PR DESCRIPTION
**Pull request by sadrasabouri**
#### Reference Issues/PRs
#28 

#### What does this implement/fix? Explain your changes.
`gc_clamp` attribute added to the `Primer` class

#### Any other comments?
:red_circle: It's odd but the test sample driven from [oligoevaluator](https://www.oligoevaluator.com/OligoCalcServlet) is different from what the issue and the references says. We should find out what is wrong with either of these sources.

### CODE REVIEW

File: opr/functions.py [hunk 1]
```diff
@@ -36,3 +38,17 @@ def basic_melting_temperature_calc(sequence):
...
     return melting_temperature
+
+
+def gc_clamp_calc(sequence):
+    """
+    Calculate GC clamp.
+
+    :param sequence: primer nucleotides sequence
+    :type sequence: str
+    :return: Number of guanine (G) or cytosine (C) bases in the last 5 bases of a primer.
```
Comment (sepandhaghighi) [inline opr/functions.py:49]: :return: <s>N</s>number of guanine (G) or cytosine (C) bases in the last 5 bases of a primer

File: opr/functions.py [hunk 2]
```diff
@@ -36,3 +38,17 @@ def basic_melting_temperature_calc(sequence):
     else:
         melting_temperature = 64.9 + 41 * ((g_count + c_count - 16.4) / (a_count + t_count + g_count + c_count))
     return melting_temperature
+
+
+def gc_clamp_calc(sequence):
+    """
+    Calculate GC clamp.
+
+    :param sequence: primer nucleotides sequence
```
Comment (sepandhaghighi) [inline opr/functions.py:47]: :param sequence: primer <s>nucleotides</s> sequence

File: opr/functions.py [hunk 3]
```diff
@@ -36,3 +38,17 @@ def basic_melting_temperature_calc(sequence):
...
+    :type sequence: str
+    :return: Number of guanine (G) or cytosine (C) bases in the last 5 bases of a primer.
+    """
+    if len(sequence) < 5:
+        warn(PRIMER_SEQUENCE_VALID_RANGE_FOR_GC_CLAMP_WARNING, RuntimeWarning)
```
Comment (sepandhaghighi) [inline opr/functions.py:52]: Remove warning

File: tests/test_calculations.py [hunk 4]
```diff
@@ -19,6 +19,14 @@ def test_gc_content_3(): #Reference: https://jamiemcgowan.ie/bioinf/gc_content.h
     oprimer = Primer("ATTTTTT")
     assert oprimer.gc_content == 0
 
+def test_gc_clamp_1(): #Reference: https://www.oligoevaluator.com/OligoCalcServlet
+    oprimer = Primer("ATCGATCGATCGATCGATCG")
```
Comment (sepandhaghighi) [inline tests/test_calculations.py:23]: - Use this sequence: `ATCGATCGATCGATCGGTCG`
- Use this website: https://www.bioinformatics.org/sms2/pcr_primer_stats.html

File: tests/test_calculations.py [hunk 5]
```diff
@@ -19,6 +19,14 @@ def test_gc_content_3(): #Reference: https://jamiemcgowan.ie/bioinf/gc_content.h
...
+    assert oprimer.gc_clamp == 2
+
+def test_gc_clamp_2(): #Reference: https://www.oligoevaluator.com/OligoCalcServlet
```
Comment (sepandhaghighi) [inline tests/test_calculations.py:26]: Use this website: https://www.bioinformatics.org/sms2/pcr_primer_stats.html

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

File: tests/test_calculations.py (hunk 4 above)
Comment (sadrasabouri) [reply tests/test_calculations.py:23]: It does not give me the `gc_clamp` value:
Screenshot from 2024-12-25 10-04-25

File: opr/functions.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 """OPR functions."""
 from .params import A_WEIGHT, T_WEIGHT, C_WEIGHT, G_WEIGHT, ANHYDROUS_MOLECULAR_WEIGHT_CONSTANT
+from warnings import warn
```
Comment (sepandhaghighi) [inline opr/functions.py:4]: Remove this line

File: tests/test_calculations.py [hunk 7]
```diff
@@ -19,6 +19,14 @@ def test_gc_content_3(): #Reference: https://jamiemcgowan.ie/bioinf/gc_content.h
...
     assert oprimer.gc_content == 0
 
+def test_gc_clamp_1(): #Reference: https://www.bioinformatics.org/sms2/pcr_primer_stats.html
+    oprimer = Primer("ATCGATCGATCGATCGGTCG")
+    assert oprimer.gc_clamp == 4
+
+def test_gc_clamp_2(): #Reference: https://www.bioinformatics.org/sms2/pcr_primer_stats.html
+    oprimer = Primer("ATCG")
+    assert oprimer.gc_clamp == 0
+
```
Comment (sepandhaghighi) [inline tests/test_calculations.py:29]: Also, please add this test case (3):

sequence: "ACTTA"

File: opr/functions.py [hunk 8]
```diff
@@ -36,3 +37,16 @@ def basic_melting_temperature_calc(sequence):
...
     return melting_temperature
+
+
+def gc_clamp_calc(sequence):
+    """
+    Calculate GC clamp.
+
+    :param sequence: primer sequence
+    :type sequence: str
+    :return: number of guanine (G) or cytosine (C) bases in the last 5 bases of a primer
```
Comment (sepandhaghighi) [inline opr/functions.py:48]: the primer

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]

**Comment by AHReccese** [comment]
@sadrasabouri @sepandhaghighi 
Guys I wanted to review this PR, but why you've removed me?

**Review by sepandhaghighi — APPROVED** [review]
🔥

**Comment by sepandhaghighi** [comment]
> @sadrasabouri @sepandhaghighi Guys I wanted to review this PR, but why you've removed me?

Hi @AHReccese 
It's not a major PR and doesn't need two reviewers.
Now you can start your work on #29.

**Pull request merged**